### PR TITLE
Add configs for Centromere keyboard

### PR DESCRIPTION
**Pull request by spe2**
## Description
Add files for the new hot-swappable Centromere keyboard.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [X] Keyboard (addition or update)
- [X] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/centromere/keymaps/default/keymap.c [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+};
+
+
+//Mousekeys
+#define MOUSEKEY_DELAY 300
+#define MOUSEKEY_INTERNAL 50
+#define MOUSEKEY_MAX_SPEED 20
+#define MOUSEKEY_TIME_TO_MAX 30
+#define MOUSEKEY_WHEEL_MAX_SPEED 8
+#define MOUSEKEY_WHEEL_TIME_TO_MAX 40
```
[the remark this answers is not in the document]
Comment (spe2) [reply keyboards/centromere/keymaps/default/keymap.c]: Suggested changes made.

File: keyboards/centromere/centromere.c [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+	PORTD |= (1<<1); // Turn it off
+	DDRC  |= (1<<4) | (1<<5); // Pins to red and blue, set as output
+	PORTC |= (1<<4) | (1<<5); // Turn them off
+
+#else
+	DDRD  |= (1<<1); // Pin to green, set as output
+	PORTD |= (1<<1); // Turn it off
+	DDRF  |= (1<<4) | (1<<5); // Pins to red and blue, set as output
+	PORTF |= (1<<4) | (1<<5); // Turn them off
+#endif
```
[the remark this answers is not in the document]
Comment (spe2) [reply keyboards/centromere/centromere.c]: @drashna Committed change to CPIO control functions.

File: keyboards/centromere/centromere.h [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+#define red_led_off   PORTF |= (1<<5)
+#define red_led_on    PORTF &= ~(1<<5)
+#define blu_led_off   PORTF |= (1<<4)
+#define blu_led_on    PORTF &= ~(1<<4)
+
+#endif
+
+#define grn_led_off   PORTD |= (1<<1)
+#define grn_led_on    PORTD &= ~(1<<1)
+
```
[the remark this answers is not in the document]
Comment (spe2) [reply keyboards/centromere/centromere.h]: Switched to GPIO control functions.

**Comment by spe2** [comment]
@drashna Suggested changes have been made.

**Comment by spe2** [comment]
@drashna  Any update on reviewing the changes made as requested?

File: keyboards/centromere/rules.mk [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+# BLUETOOTH_ENABLE = yes # Enable Bluetooth with the Adafruit EZ-Key HID
+
+USB = /dev/ttyACM0
+
+# upload: build
+# 	$(CENTROMERE_UPLOAD_COMMAND)
+
+OPT_DEFS += -DCENTROMERE_PROMICRO
+CENTROMERE_UPLOAD_COMMAND = while [ ! -r $(USB) ]; do sleep 1; done; \
+                         avrdude -p $(MCU) -c avr109 -U flash:w:$(TARGET).hex -P $(USB)
```
Comment (drashna) [inline keyboards/centromere/rules.mk]: I'm not sure that this is needed.  The `:avrdude` target should handle this just fine, on it's own.

File: keyboards/centromere/matrix.c [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+    {
+        //shifting and transferring the keystates to the QMK matrix variable
+        for (uint8_t i = 0; i < MATRIX_ROWS; i++) {
+            matrix[i] = (uint16_t) uart_data[i*2] | (uint16_t) uart_data[i*2+1] << 5;
+        }
+    }
+
+
+    matrix_scan_quantum();
+    return 1;
```
Comment (drashna) [inline keyboards/centromere/matrix.c]: I'm not exactly sure how you'd implement it, but ideally, this should only return a value if the matrix has changed. 

However, it should be okay for now.

File: keyboards/centromere/rules.mk (hunk 4 above)
Comment (spe2) [reply keyboards/centromere/rules.mk]: @drashna You're correct, it seems to work just fine without.

File: keyboards/centromere/rules.mk [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+#   ATmega32A    bootloadHID
+#   ATmega328P   USBasp
+BOOTLOADER = caterina
+F_CPU = 8000000
+F_USB = $(F_CPU)
+
+# Build Options
+#   comment out to disable the options.
+#
+#BOOTMAGIC_ENABLE = yes	# Virtual DIP switch configuration(+1000)
```
Comment (fauxpark) [inline keyboards/centromere/rules.mk]:
```suggestion
BOOTMAGIC_ENABLE = no	# Virtual DIP switch configuration
```
Yuo may want to make this `lite` instead: https://docs.qmk.fm/#/feature_bootmagic?id=bootmagic-lite

File: keyboards/centromere/rules.mk [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+#   Teensy       halfkay
+#   Pro Micro    caterina
+#   Atmel DFU    atmel-dfu
+#   LUFA DFU     lufa-dfu
+#   QMK DFU      qmk-dfu
+#   ATmega32A    bootloadHID
+#   ATmega328P   USBasp
+BOOTLOADER = caterina
+F_CPU = 8000000
+F_USB = $(F_CPU)
```
Comment (fauxpark) [inline keyboards/centromere/rules.mk]:
```suggestion
```

File: keyboards/centromere/rules.mk [hunk 8]
```diff
@@ -0,0 +1,41 @@
+# MCU name
+MCU = atmega32u4
+
```
Comment (fauxpark) [inline keyboards/centromere/rules.mk]:
```suggestion

# Processor frequency
F_CPU = 8000000

```

File: keyboards/centromere/rules.mk [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+# Bootloader selection
+#   Teensy       halfkay
+#   Pro Micro    caterina
+#   Atmel DFU    atmel-dfu
+#   LUFA DFU     lufa-dfu
+#   QMK DFU      qmk-dfu
+#   ATmega32A    bootloadHID
+#   ATmega328P   USBasp
+BOOTLOADER = caterina
+F_CPU = 8000000
```
Comment (fauxpark) [inline keyboards/centromere/rules.mk]:
```suggestion
```

File: keyboards/centromere/rules.mk [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+F_USB = $(F_CPU)
+
+# Build Options
+#   comment out to disable the options.
+#
+#BOOTMAGIC_ENABLE = yes	# Virtual DIP switch configuration(+1000)
+MOUSEKEY_ENABLE = yes	# Mouse keys(+4700)
```
Comment (fauxpark) [inline keyboards/centromere/rules.mk]:
```suggestion
MOUSEKEY_ENABLE = yes	# Mouse keys
```

File: keyboards/centromere/rules.mk [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+EXTRAKEY_ENABLE = yes	# Audio control and System control(+450)
```
Comment (fauxpark) [inline keyboards/centromere/rules.mk]:
```suggestion
EXTRAKEY_ENABLE = yes	# Audio control and System control
```

File: keyboards/centromere/rules.mk [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+CONSOLE_ENABLE = yes	# Console for debug(+400)
```
Comment (fauxpark) [inline keyboards/centromere/rules.mk]:
```suggestion
CONSOLE_ENABLE = yes	# Console for debug
```

File: keyboards/centromere/rules.mk [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+COMMAND_ENABLE = yes   # Commands for debug and configuration
+CUSTOM_MATRIX = yes    # Remote matrix from the wireless bridge
+# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
+# SLEEP_LED_ENABLE = yes  # Breathing sleep LED during USB suspend
+NKRO_ENABLE = yes		# USB Nkey Rollover - not yet supported in LUFA
+# BACKLIGHT_ENABLE = yes  # Enable keyboard backlight functionality
+# MIDI_ENABLE = yes 		# MIDI controls
+UNICODE_ENABLE = yes 		# Unicode
+# BLUETOOTH_ENABLE = yes # Enable Bluetooth with the Adafruit EZ-Key HID
```
Comment (fauxpark) [inline keyboards/centromere/rules.mk]:
```suggestion
BLUETOOTH_ENABLE = no # Enable Bluetooth with the Adafruit EZ-Key HID
```

File: keyboards/centromere/rules.mk [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+MOUSEKEY_ENABLE = yes	# Mouse keys(+4700)
+EXTRAKEY_ENABLE = yes	# Audio control and System control(+450)
+CONSOLE_ENABLE = yes	# Console for debug(+400)
+COMMAND_ENABLE = yes   # Commands for debug and configuration
+CUSTOM_MATRIX = yes    # Remote matrix from the wireless bridge
+# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
+# SLEEP_LED_ENABLE = yes  # Breathing sleep LED during USB suspend
+NKRO_ENABLE = yes		# USB Nkey Rollover - not yet supported in LUFA
+# BACKLIGHT_ENABLE = yes  # Enable keyboard backlight functionality
+# MIDI_ENABLE = yes 		# MIDI controls
```
Comment (fauxpark) [inline keyboards/centromere/rules.mk]:
```suggestion
MIDI_ENABLE = no 		# MIDI controls
```

File: keyboards/centromere/rules.mk [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+#BOOTMAGIC_ENABLE = yes	# Virtual DIP switch configuration(+1000)
+MOUSEKEY_ENABLE = yes	# Mouse keys(+4700)
+EXTRAKEY_ENABLE = yes	# Audio control and System control(+450)
+CONSOLE_ENABLE = yes	# Console for debug(+400)
+COMMAND_ENABLE = yes   # Commands for debug and configuration
+CUSTOM_MATRIX = yes    # Remote matrix from the wireless bridge
+# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
+# SLEEP_LED_ENABLE = yes  # Breathing sleep LED during USB suspend
+NKRO_ENABLE = yes		# USB Nkey Rollover - not yet supported in LUFA
+# BACKLIGHT_ENABLE = yes  # Enable keyboard backlight functionality
```
Comment (fauxpark) [inline keyboards/centromere/rules.mk]:
```suggestion
BACKLIGHT_ENABLE = no  # Enable keyboard backlight functionality
```

File: keyboards/centromere/rules.mk [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+#   comment out to disable the options.
+#
+#BOOTMAGIC_ENABLE = yes	# Virtual DIP switch configuration(+1000)
+MOUSEKEY_ENABLE = yes	# Mouse keys(+4700)
+EXTRAKEY_ENABLE = yes	# Audio control and System control(+450)
+CONSOLE_ENABLE = yes	# Console for debug(+400)
+COMMAND_ENABLE = yes   # Commands for debug and configuration
+CUSTOM_MATRIX = yes    # Remote matrix from the wireless bridge
+# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
+# SLEEP_LED_ENABLE = yes  # Breathing sleep LED during USB suspend
```
Comment (fauxpark) [inline keyboards/centromere/rules.mk]:
```suggestion
SLEEP_LED_ENABLE = no  # Breathing sleep LED during USB suspend
```

File: keyboards/centromere/rules.mk [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+NKRO_ENABLE = yes		# USB Nkey Rollover - not yet supported in LUFA
```
Comment (fauxpark) [inline keyboards/centromere/rules.mk]:
```suggestion
NKRO_ENABLE = yes		# USB Nkey Rollover
```

File: keyboards/centromere/rules.mk [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+# MIDI_ENABLE = yes 		# MIDI controls
+UNICODE_ENABLE = yes 		# Unicode
+# BLUETOOTH_ENABLE = yes # Enable Bluetooth with the Adafruit EZ-Key HID
+
+USB = /dev/ttyACM0
+
+# upload: build
+# 	$(CENTROMERE_UPLOAD_COMMAND)
+
+OPT_DEFS += -DCENTROMERE_PROMICRO
```
Comment (fauxpark) [inline keyboards/centromere/rules.mk]: This can probably be deleted.

File: keyboards/centromere/keymaps/default_u2/rules.mk [hunk 19]
```diff
@@ -0,0 +1,3 @@
+MCU = atmega32u2
```
Comment (fauxpark) [inline keyboards/centromere/keymaps/default_u2/rules.mk]:
```suggestion
# MCU name
MCU = atmega32u2
```

File: keyboards/centromere/keymaps/default_u2/rules.mk [hunk 20]
```diff
@@ -0,0 +1,3 @@
...
+BOOTLOADER = atmel-dfu
```
Comment (fauxpark) [inline keyboards/centromere/keymaps/default_u2/rules.mk]:
```suggestion

# Bootloader selection
#   Teensy       halfkay
#   Pro Micro    caterina
#   Atmel DFU    atmel-dfu
#   LUFA DFU     lufa-dfu
#   QMK DFU      qmk-dfu
#   ATmega32A    bootloadHID
#   ATmega328P   USBasp
BOOTLOADER = atmel-dfu
```

File: keyboards/centromere/keymaps/default_u2/rules.mk [hunk 21]
```diff
@@ -0,0 +1,3 @@
...
+#define CPU_TYPE ATMEGA32U2
```
Comment (fauxpark) [inline keyboards/centromere/keymaps/default_u2/rules.mk]:
```suggestion
```

File: keyboards/centromere/keymaps/default_u2/keymap.c [hunk 22]
```diff
@@ -0,0 +1,69 @@
...
+
+  [_STD] = LAYOUT( \
+  //,-----------------------------------------------------.                    ,-----------------------------------------------------.
+       KC_TAB,    KC_Q,    KC_W,    KC_E,    KC_R,    KC_T,                         KC_Y,    KC_U,    KC_I,    KC_O,   KC_P,  KC_BSPC,\
+  //|--------+--------+--------+--------+--------+--------|                    |--------+--------+--------+--------+--------+--------|
+      KC_LCTL,    KC_A,    KC_S,    KC_D,    KC_F,    KC_G,                         KC_H,    KC_J,    KC_K,    KC_L, KC_SCLN, KC_QUOT,\
+  //|--------+--------+--------+--------+--------+--------|                    |--------+--------+--------+--------+--------+--------|
+      KC_LSFT,    KC_Z,    KC_X,    KC_C,    KC_V,    KC_B,                         KC_N,    KC_M, KC_COMM,  KC_DOT, KC_SLSH,  KC_ENT,\
+  //|--------+--------+--------+--------+--------+--------+--------|  |--------+--------+--------+--------+--------+--------+--------|
+                                          KC_LGUI, MO(_FN),  KC_SPC,     KC_SPC,MO(_NUM),  KC_RALT \
```
Comment (fauxpark) [inline keyboards/centromere/keymaps/default_u2/keymap.c]: The backslashes at the end of these lines are not necessary. You can keep them if you want as it isn't hurting anything, but IMO it looks neater without them.

File: keyboards/centromere/keymaps/default/keymap.c [hunk 23]
```diff
@@ -0,0 +1,70 @@
...
+      KC_LSFT, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX,                      KC_UNDS, KC_PLUS, KC_LBRC, KC_RBRC, KC_BSLS, KC_TILD,\
+
+                                          KC_LGUI, MO(_FN),  KC_SPC,     KC_ENT,MO(_NUM), KC_RALT \
+                                      //`--------------------------'  `--------------------------'
+  ),
+
+};
+
+void matrix_scan_user(void) {
+    uint8_t layer = biton32(layer_state);
```
Comment (fauxpark) [inline keyboards/centromere/keymaps/default/keymap.c]:
```suggestion
    uint8_t layer = get_highest_layer(layer_state);
```

File: keyboards/centromere/keymaps/default_u2/keymap.c [hunk 24]
```diff
@@ -0,0 +1,69 @@
...
+      KC_LSFT, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX,                      KC_UNDS, KC_PLUS, KC_LBRC, KC_RBRC, KC_BSLS, KC_TILD,\
+
+                                          KC_LGUI, MO(_FN),  KC_SPC,     KC_ENT,MO(_NUM), KC_RALT \
+                                      //`--------------------------'  `--------------------------'
+  ),
+
+};
+
+void matrix_scan_user(void) {
+    uint8_t layer = biton32(layer_state);
```
Comment (fauxpark) [inline keyboards/centromere/keymaps/default_u2/keymap.c]:
```suggestion
    uint8_t layer = get_highest_layer(layer_state);
```

File: keyboards/centromere/centromere.h [hunk 25]
```diff
@@ -0,0 +1,49 @@
+#ifndef CENTROMERE_H
```
Comment (fauxpark) [inline keyboards/centromere/centromere.h]:
```suggestion
#pragma once
```

File: keyboards/centromere/centromere.h [hunk 26]
```diff
@@ -0,0 +1,49 @@
...
+#define CENTROMERE_H
```
Comment (fauxpark) [inline keyboards/centromere/centromere.h]:
```suggestion
```

File: keyboards/centromere/centromere.h [hunk 27]
```diff
@@ -0,0 +1,49 @@
...
+                 k32, k33, k34,      k35, k36, k37  \
+) \
+  {                                                           \
+    { k00,     k01, k02, k03, k04, k05, k06, k07, k08,   k09   }, \
+    { k10,     k11, k12, k13, k14, k15, k16, k17, k18,   k19   }, \
+    { k20,     k21, k22, k23, k24, k25, k26, k27, k28,   k29   }, \
+    { KC_NO, KC_NO, k32, k33, k34, k35, k36, k37, KC_NO, KC_NO }, \
+    { KC_NO, KC_NO, k2a, k1a, k0a, k0b, k1b, k2b, KC_NO, KC_NO }  \
+  }
+#endif
```
Comment (fauxpark) [inline keyboards/centromere/centromere.h]:
```suggestion
```

File: keyboards/centromere/centromere.c [hunk 28]
```diff
@@ -0,0 +1,48 @@
+#include "centromere.h"
+
+void uart_init(void) {
+	SERIAL_UART_INIT();
+}
+
+void led_init(void) {
+#if CPU_TYPE == ATMEGA32U2
```
Comment (fauxpark) [inline keyboards/centromere/centromere.c]:
```suggestion
#if MCU == atmega32u2
```

File: keyboards/centromere/centromere.h [hunk 29]
```diff
@@ -0,0 +1,47 @@
+#pragma once
+
+#include "quantum.h"
+
+#if CPU_TYPE == ATMEGA32U2
```
Comment (fauxpark) [inline keyboards/centromere/centromere.h]:
```suggestion
#if MCU = atmega32u2
```

File: keyboards/centromere/centromere.h [hunk 30]
```diff
@@ -0,0 +1,47 @@
+#pragma once
+
+#include "quantum.h"
+
+#if MCU = atmega32u2
```
Comment (fauxpark) [inline keyboards/centromere/centromere.h]:
```suggestion
#if MCU == atmega32u2
```

Sorry 🤦‍♂

**Comment by spe2** [comment]
Hi there @drashna , have you had  a chance to look at the changes to merge this request?

**Comment by skorokithakis** [comment]
I would like to add my support for this PR here, as there are some features I'm missing (mainly KC_BRIU/KC_BRID).

**Comment by spe2** [comment]
@fauxpark @drashna Is there anything else that needs to be done to move this forward?

**Comment by spe2** [comment]
@fauxpark  I'd prefer to have this integrated and work on the custom matrix lite code in a followup PR.

**Comment by skorokithakis** [comment]
Can this be merged? I have a Centromere Plus I'd like to be able to use, and this has been approved.

**Comment by skorokithakis** [comment]
Thanks @drashna! @spe2, this PR includes the config for the Centromere Plus v2, right?

**Comment by noroadsleft** [comment]
@skorokithakis Near as I can tell, this only supports the non-Plus Centromere. The Plus has a different layout.

Hopefully there will be a PR for the Plus soon.

**Comment by skorokithakis** [comment]
Oh :( I hope so! @spe2, do you have any plans for that?

**Comment by skorokithakis** [comment]
@spe2 any update on the Plus?

**Pull request merged**